libbuteosyncfw/clientfw/SyncClientInterface.h

```diff
@@ -182,14 +205,53 @@ class SyncClientInterface: public QObject
      */
     QStringList syncProfilesByKey(const QString &aKey, const QString &aValue);
 
+    /*! \brief asynchronous version of syncProfilesByKey().
+     *
+     * \param aKey Key to match for profile.
+     * \param aValue Value to match for profile.
+     * \return a newly created watcher on a QDBusPendingReply<QStringList>.
+     */
+    QDBusPendingCallWatcher* requestSyncProfilesByKey(const QString &aKey, const QString &aValue) const;
+
+    /*! \brief Gets profiles matching the profile type.
+     *
+     * \param aType Type of the profile service/storage/sync.
+     * \return The profiles as Xml string list.
+     */
+    QStringList profilesByType(const QString &aType);
+
+    /*!
+     * \brief asynchronous version of profilesByType()
+     *
+     * \param aType Type of the profile service/storage/sync.
+     * \return a newly created watcher on a QDBusPendingReply<QStringList>.
+     */
+    QDBusPendingCallWatcher* requestProfilesByType(const QString &aProfileId) const;
+
     /*! \brief Gets a profiles  matching the profile type.
      *
      * \param aType Type of the profile service/storage/sync.
      * \return The sync profile ids as string list.
      */
     QStringList syncProfilesByType(const QString &aType);
+
+    /*!
+     * \brief creates a process singleton
+     *
+     * Use this instance to share a same D-Bus connection to the daemon
+     * between several clients in a process.
+     * \return a singleton instance, don't free it.
+     */
+    static SyncClientInterface* instance();
```

Comment:
One problem with this sort of singleton api is that it will leak memory. 

It could be avoided internally by using Q_GLOBAL_STATIC for creating the actual instance, but that might have problems if qdbus stuff is deleted after the qcoreapplication and qt stuff is already gone.

Another alternative would be to return a QSharedPointer, which would actually even delete the dbus stuff if anyone is not using it anymore. See e.g.
https://github.com/sailfishos/libconnman-qt/blob/master/libconnman-qt/networkmanager.cpp#L649

Reply:
Thank you for the example in libconnman-qt. I've done the same to use a QSharedPointer. Initially, I chose the pointer solution because it's not really a memory leak, just that it's unfreed memory on quit. Or a fixed amount of memory leak. A second because, if the QML is quickly creating and destroying objects using a SyncClientInterface, then the instance stays alive and is reused. But I guess the QSharedPointer approach is cleaner, for no real gain.

Reply:
Granted that it's a bit more controlled 'leak', but anyhow good to avoid if easily possible. Could show up as false positive or on worse case the missing dtor might some day have functionality that was expected executed. Thanks for changes.

Reply:
Good point about the destructor that may one day contains necessary actions. Thanks for the explanation.